### PR TITLE
*: remove snapshot generation V1 API

### PR DESCRIPTION
**Pull request by choleraehyq**
- [x] remove all snap v1 api and tests
- [x] tidy v2 code (move them out of mod v2 etc.)
- [x] performance test and optimization.

We get some flames graph to do optimization, and finally we found out that we need to wait for this https://github.com/pingcap/tikv/pull/1983 to do deeper optimization. Will do in another PR.

~~This PR will break compatibility, but we haven't made a dicision whether we will provide compatibility to v1 snapshot file.~~
Finally we decide not to provide compatibility to v1 snapshot file. You need to reimport data to upgrade to this version.

We should make this PR merge meet Pre-GA milestone, so we will only remove V1 in this PR.

Signed-off-by: Cholerae Hu <huyingqian@pingcap.com>

### CODE REVIEW

File: src/raftstore/store/snap.rs [hunk 1]
```diff
@@ -195,1874 +199,721 @@ pub fn retry_delete_snapshot(deleter: Box<SnapshotDeleter>,
...
-            info!("[region {}] scan snapshot {}, size {}, key count {}, takes {:?}",
-                  region.get_id(),
-                  self.path(),
-                  size,
-                  stat.kv_count,
-                  t.elapsed());
-
-            Ok(())
-        }
+    pub fn new_for_building<T: Into<PathBuf>>(dir: T,
```
Comment (zhangjinpeng1987) [inline src/raftstore/store/snap.rs]: Can we discard these `new_for_xxxx` functions?

Comment (choleraehyq) [inline src/raftstore/store/snap.rs]: I'm considering about this. Any advices?

File: src/raftstore/store/peer_storage.rs [hunk 2]
```diff
@@ -1641,10 +1626,12 @@ mod test {
         assert_eq!(s1.truncated_index(), 3);
         assert_eq!(s1.truncated_term(), 3);
 
-        let key = SnapKey::from_snap(&snap1).unwrap();
-        let from = mgr.get_snapshot_for_sending(&key).unwrap();
-        let to = mgr.get_snapshot_for_receiving(&key, b"").unwrap();
-        copy_snapshot(from, to).unwrap();
+        //        What does these test mean??
```
Comment (siddontang) [inline src/raftstore/store/peer_storage.rs]: why comment these?

**Comment by siddontang** [comment]
because this PR is not compatibility, I will rebase it to the pre-ga branch.

**Comment by siddontang** [comment]
I suggest removing only v1 snapshot in this PR, and then do other things later.

File: src/raftstore/store/peer_storage.rs (hunk 2 above)
Comment (choleraehyq) [inline src/raftstore/store/peer_storage.rs]: These tests is used to test v1 snapshot.

**Comment by choleraehyq** [comment]
@hhkbp2 ptal

File: src/coprocessor/codec/mysql/json/mod.rs [hunk 3]
```diff
@@ -47,3 +53,77 @@ pub enum Json {
     Boolean(bool),
     None,
 }
+
+// https://dev.mysql.com/doc/refman/5.7/en/json-creation-functions.html#function_json-array
+pub fn json_array(elems: Vec<Datum>) -> Result<Json> {
```
Comment (zhangjinpeng1987) [inline src/coprocessor/codec/mysql/json/mod.rs]: Why this pr contain json?

Comment (choleraehyq) [inline src/coprocessor/codec/mysql/json/mod.rs]: Wooo, I rebase my branch with master, and there are some commits that have been merged into master but haven't been merged in pre-ga

Comment (choleraehyq) [inline src/coprocessor/codec/mysql/json/mod.rs]: merge master into pre-ga may solve this.

File: src/raftstore/store/snap.rs [hunk 4]
```diff
@@ -2209,30 +1135,21 @@ impl SnapManager {
...
-            if s.exists() {
-                return Ok(Box::new(s));
-            }
-        }
-        let s = try!(v1::Snap::new_for_reading(&core.base, core.snap_size.clone(), false, key));
+        let s = Snap::new_for_applying(&core.base,
+                                       key,
+                                       core.snap_size.clone(),
+                                       Box::new(self.clone()))?;
+        assert!(s.exists());
```
Comment (zhangjinpeng1987) [inline src/raftstore/store/snap.rs]: Can we just return an error when snapshot not exist?

**Comment by zhangjinpeng1987** [comment]
LGTM
@BusyJay @hhkbp2 PTAL

File: src/raftstore/store/peer_storage.rs [hunk 5]
```diff
@@ -1329,25 +1329,18 @@ mod test {
     }
 
     #[test]
-    fn test_storage_create_snapshot_with_plain_format() {
-        test_storage_create_snapshot(false);
-    }
-
-    #[test]
     fn test_storage_create_snapshot_with_sst_format() {
```
Comment (hhkbp2) [inline src/raftstore/store/peer_storage.rs]: Remove this test case. Use `test_storage_create_snapshot()` directly.

File: src/raftstore/store/cmd_resp.rs [hunk 6]
```diff
@@ -11,7 +11,6 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
-use std::boxed::Box;
```
Comment (hhkbp2) [inline src/raftstore/store/cmd_resp.rs]: Is this line of code relevant?

File: tests/raftstore/pd.rs [hunk 7]
```diff
@@ -13,7 +13,6 @@
 
 
 use std::collections::{HashMap, BTreeMap, HashSet};
-use std::vec::Vec;
```
Comment (hhkbp2) [inline tests/raftstore/pd.rs]: Is this line of code relevant?

File: src/raftstore/store/snap.rs [hunk 8]
```diff
@@ -2493,64 +1938,8 @@ mod test {
...
-        s4.apply(options).unwrap();
-
-        v2::test::assert_eq_db(db, dst_db.as_ref());
-    }
-
-    #[test]
     fn test_snap_deletion_on_registry() {
-        test_snap_deletion_on_registry_impl(false);
-        test_snap_deletion_on_registry_impl(true);
+        test_snap_deletion_on_registry_impl();
```
Comment (hhkbp2) [inline src/raftstore/store/snap.rs]: Rename `test_snap_deletion_on_registry_impl()` to `test_snap_deletion_on_registry()` directly.

**Comment by hhkbp2** [comment]
Rest LGTM

File: src/raftstore/store/snap.rs (hunk 1 above)
Comment (siddontang) [inline src/raftstore/store/snap.rs]: Consider it in next PR, not here.

**Comment by siddontang** [comment]
This PR is too huge and hard to review, please split it into small PRs.

1. Remove V1 and still keep V2
2. Move V2 to super
3. Tiny Clean

**Comment by choleraehyq** [comment]
Ok, close this.

**Pull request closed without merging**